device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/fileupload/FileUploadTask.java

```diff
@@ -199,9 +208,31 @@ public IotHubStatusCode sendNotification(FileUploadCompletionNotification fileUp
         ResponseMessage responseMessage = httpsTransportManager.sendFileUploadNotification(message);
         httpsTransportManager.close();
 
+        validateServiceStatusCode(responseMessage, "Failed to complete the file upload notification");
+
         return responseMessage.getStatus();
     }
 
+    private void validateServiceStatusCode(ResponseMessage responseMessage, String errorMessage) throws IOException
+    {
+        String responseMessagePayload = null;
+        if (responseMessage.getBytes() != null && responseMessage.getBytes().length > 0)
+        {
+            responseMessagePayload = new String(responseMessage.getBytes(), DEFAULT_IOTHUB_MESSAGE_CHARSET);
+        }
+
+        IotHubServiceException serviceException =
+            IotHubStatusCode.getConnectionStatusException(
+                responseMessage.getStatus(),
+                responseMessagePayload);
+
+        // serviceException is only not null if the provided status code was a non-successful status code like 400, 429, 500, etc.
+        if (serviceException != null)
+        {
+            throw new IOException(errorMessage, serviceException);
```

Comment:
Is it right to throw IOException here? I understand you may not want to change behavior but shouldn't it be wrapped in IoTHub Exception type?

Reply:
You have it backwards. In order to not cause a breaking change, the IoTHubServiceException needs to be the nested exception within an IOException